<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=11"/>
<meta name="generator" content="Doxygen 1.9.4"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>Legofit: daf</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr id="projectrow">
  <td id="projectalign">
   <div id="projectname">Legofit
   </div>
   <div id="projectbrief">infers population history from nucleotide site patterns.</div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.9.4 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
var searchBox = new SearchBox("searchBox", "search",'Search','.html');
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
/* @license-end */
</script>
<div id="main-nav"></div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

</div><!-- top -->
<div><div class="header">
  <div class="headertitle"><div class="title">daf format </div></div>
</div><!--header-->
<div class="contents">
<div class="textblock"><p >Calculate derived allele frequency, daf.</p>
<p >Input file should consist of tab-separated columns:</p>
<ol type="1">
<li>chromosome</li>
<li>position</li>
<li>reference allele</li>
<li>alternate alleles</li>
<li>ancestral allele</li>
<li>genotype in format "0/1" or "0|1", where 0 represents a copy of the reference allele and 1 a copy of the derived allele.</li>
<li>etc for as many columns as there are genotypes.</li>
</ol>
<p >This can be generated from a vcf file that includes annotations for ancestral alleles. If the ancestral is labelled "AA", the input for daf can be generated, using bcftools, as follows:</p>
<p >bcftools query -f 'CHROM\tPOS\tREF\tALT\tINFO/AA[\tGT]<br  />
' fname.vcf.gz</p>
<p >Output is in 5 columns, separated by whitespace:</p>
<ol type="1">
<li>chromosome</li>
<li>position of the nucleotide</li>
<li>aa, the ancestral allele</li>
<li>da, the derived allele</li>
<li>daf, derived allele frequency</li>
</ol>
<p >The input files should include all sites at which derived alleles are present in any of the populations under study. For example, consider an analysis involving modern humans and Neanderthals. The modern human data must include all sites at which Neanderthals carry derived alleles, even if these sites do not vary among modern humans. To accomplish this, it is best to use whole-genome data for all populations.</p>
<p >The input should not contain duplicate nucleotide sites, the chromosomes should be sorted in lexical order, and within each chromosome, the nucleotides should be in numerical order. Otherwise, daf will abort with an error.</p>
<p >Sites are rejected unless they have a single ref or ancestral allele. Missing values are allowed for the alt allele. At the end of the job a summary of rejected sites is written to stderr.</p>
<dl class="section copyright"><dt>Copyright</dt><dd>Copyright (c) 2016, Alan R. Rogers <a href="#" onclick="location.href='mai'+'lto:'+'rog'+'er'+'s@a'+'nt'+'hro'+'.u'+'tah'+'.e'+'du'; return false;">roger<span class="obfuscator">.nosp@m.</span>s@an<span class="obfuscator">.nosp@m.</span>thro.<span class="obfuscator">.nosp@m.</span>utah<span class="obfuscator">.nosp@m.</span>.edu</a>. This file is released under the Internet Systems Consortium License, which can be found in file "LICENSE".</dd></dl>
<p>Input file should consist of tab-separated columns:</p>
<ol type="1">
<li>chromosome</li>
<li>position</li>
<li>reference allele</li>
<li>alternate alleles</li>
<li>ancestral allele</li>
<li>genotype in format "0/1" or "0|1", where 0 represents a copy of the reference allele and 1 a copy of the derived allele.</li>
<li>etc for as many columns as there are genotypes.</li>
</ol>
<p >This can be generated from a vcf file that includes annotations for ancestral alleles. If the ancestral is labelled "AA", the input for daf can be generated, using bcftools, as follows:</p>
<p >bcftools query -f 'CHROM\tPOS\tREF\tALT\tINFO/AA[\tGT]<br  />
' fname.vcf.gz</p>
<p >Output is in 5 columns, separated by whitespace:</p>
<ol type="1">
<li>chromosome</li>
<li>position of the nucleotide</li>
<li>aa, the ancestral allele</li>
<li>da, the derived allele</li>
<li>daf, derived allele frequency</li>
</ol>
<p >The input files should include all sites at which derived alleles are present in any of the populations under study. For example, consider an analysis involving modern humans and Neanderthals. The modern human data must include all sites at which Neanderthals carry derived alleles, even if these sites do not vary among modern humans. To accomplish this, it is best to use whole-genome data for all populations.</p>
<p >The input should not contain duplicate nucleotide sites, the chromosomes should be sorted in lexical order, and within each chromosome, the nucleotides should be in numerical order. Otherwise, daf will abort with an error.</p>
<p >Sites are rejected unless they have a single ref or ancestral allele. Missing values are allowed for the alt allele. At the end of the job a summary of rejected sites is written to stderr.</p>
<dl class="section note"><dt>Note</dt><dd>Ryan Bohlender 10-6-2017 Assuming a file sorted by position, within each chromosome, we can compare duplicates as they will always be adjacent. Given a qual value strictly greater than the previous line, we will print the current line, otherwise the previous line. To handle multiple duplicates, we will output the line upon reading a new position, or EOF.</dd></dl>
<dl class="section copyright"><dt>Copyright</dt><dd>Copyright (c) 2016, Alan R. Rogers <a href="#" onclick="location.href='mai'+'lto:'+'rog'+'er'+'s@a'+'nt'+'hro'+'.u'+'tah'+'.e'+'du'; return false;">roger<span class="obfuscator">.nosp@m.</span>s@an<span class="obfuscator">.nosp@m.</span>thro.<span class="obfuscator">.nosp@m.</span>utah<span class="obfuscator">.nosp@m.</span>.edu</a>. This file is released under the Internet Systems Consortium License, which can be found in file "LICENSE". </dd></dl>
</div></div><!-- contents -->
</div><!-- PageDoc -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated on Sat Feb 25 2023 07:43:34 for Legofit by&#160;<a href="https://www.doxygen.org/index.html"><img class="footer" src="doxygen.svg" width="104" height="31" alt="doxygen"/></a> 1.9.4
</small></address>
</body>
</html>
