<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" lang="en-US">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=11"/>
<meta name="generator" content="Doxygen 1.14.0"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>Legofit: legosim</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<script type="text/javascript" src="clipboard.js"></script>
<script type="text/javascript" src="cookie.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr id="projectrow">
  <td id="projectalign">
   <div id="projectname">Legofit
   </div>
   <div id="projectbrief">infers population history from nucleotide site patterns.</div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.14.0 -->
<script type="text/javascript">
var searchBox = new SearchBox("searchBox", "search/",'.html');
</script>
<script type="text/javascript">
$(function() { codefold.init(); });
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
$(function() {
  initMenu('',true,false,'search.php','Search',false);
  $(function() { init_search(); });
});
</script>
<div id="main-nav"></div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<div id="MSearchResults">
<div class="SRPage">
<div id="SRIndex">
<div id="SRResults"></div>
<div class="SRStatus" id="Loading">Loading...</div>
<div class="SRStatus" id="Searching">Searching...</div>
<div class="SRStatus" id="NoMatches">No Matches</div>
</div>
</div>
</div>
</div>

</div><!-- top -->
<div id="doc-content">
<div><div class="header">
  <div class="headertitle"><div class="title">legosim </div></div>
</div><!--header-->
<div class="contents">
<div class="textblock"><p>Generate site patterns by coalescent simulation</p>
<h1 class="doxsection"><a class="anchor" id="autotoc_md10"></a>
legosim: coalescent simulations within a network of populations</h1>
<pre class="fragment">usage: legosim [options] input_file_name
   where options may include:
   -b or --branch_length
      print branch lengths rather than probabilities
   -i &lt;x&gt; or --nItr &lt;x&gt;
      number of iterations in simulation
   -1 or --singletons
      Use singleton site patterns
   -d &lt;x&gt; or --deterministic &lt;x&gt;
      Deterministic algorithm, ignoring states with Pr &lt;= x
   --network
      Print summary of population network and exit
   --plot &lt;filename&gt; or -p &lt;filename&gt;
      Create a dot-format file for plotting the network.
   -U &lt;x&gt;
      Mutations per generation per haploid genome.
   -h or --help
      print this message
   --version
      print version and exit
Options -i, --nItr, and -U cannot be used with --deterministic.
Option -U cannot be used with -b or --branch_length.
</pre><p>Here, "input_file" should be in <a class="el" href="index.html#lgo">.lgo</a> format, which describes the history of population size, subdivision, and gene flow. The output looks like this: </p><pre class="fragment">#######################################
# legosim: site pattern probabilities #
#   version 2.0.1-12-gf241700-dirty   #
#######################################

# Program was compiled: Nov 21 2020 10:15:39
# Program was run: Sat Nov 21 10:17:57 2020

# cmd: ./legosim -1 -d 0 input.lgo
# input file                  : input.lgo
# algorithm                   : deterministic
# ignoring probs &lt;=           : 0
# including singleton site patterns.
#       SitePat            Prob
              x    0.1560242402
              y    0.1481456533
              n    0.4023561523
            x:y    0.2854101359
            x:n    0.0000926157
            y:n    0.0079712026
</pre><p>Here, the "SitePat" column labels site patterns. For example, site pattern xy (denoted by "x:y" in this output) refers to nocleotide sites at which the derived allele is present in single haploid samples from X and Y but not in samples from other populations. The probability of this site pattern, under the model of history specified in file <span class="tt">input.lgo</span>, is given in the "Prob" column. When the <span class="tt">-d</span> or <span class="tt">--deterministic</span> options are used, these probabilities are calculated by a deterministic algorithm.</p>
<p>In the example above, I used <span class="tt">-d 0</span>, which tells legosim to ignore only those states with probability 0. Thus, the answers will be extremely accurate. When neither <span class="tt">-d</span> nor <span class="tt">--deterministic</span> are used, the probabilities are estimated by coalescent simulation.</p>
<h2 class="doxsection"><a class="anchor" id="autotoc_md11"></a>
Ignoring states of low probability</h2>
<p>The deterministic algorithm sums across all possible histories of the samples defined in the .lgo file. The number of histories increases rapidly with sample size and with the number of migration events. For larger models, it is not feasible to use <span class="tt">-d 0</span>. However, one can still use the deterministic algorithm to obtain an approximate answer, using an argument such as <span class="tt">-d 1e-6</span>. This tells legosim to use the deterministic algorithm while ignoring states whose probability is less than or equal to 1e-6. Nonzero arguments to <span class="tt">-d</span> will introduce bias. To evaluate the magnitude of this bias, compare the results to those obtained using the stochastic algorithm or (where feasible) the deterministic one with <span class="tt">-d 0</span>.</p>
<p>When the deterministic algorithm is used, legosim prints a line of output that looks like this: </p><pre class="fragment"># Summed probability of ignored improbable events: 3.515617e-04
</pre><p>This "summed probability" can exceed 1, because it sums across several probability distributions. Nonetheless, when it is small, that is an indication that we have not ignored much of any of these distributions. To rexpress this as a fraction of the of the total probability mass, run legosim with argument <span class="tt">-d 1</span>. This ignores all events, so the summed probability of ignored events will be an integer, which equals the number of probability distributions involved. Divide the original summed probability by this number to express it as a fraction of total probability mass.</p>
<h2 class="doxsection"><a class="anchor" id="autotoc_md12"></a>
Simulating site pattern counts</h2>
<p>To simulate site pattern counts across an entire genome, use the <span class="tt">-U</span> option, whose argument give the expected number of mutations per generation per haploid genome. This argument should equal <picture><source srcset="form_4_dark.png" media="(prefers-color-scheme: dark)"/><img class="formulaInl" alt="$\mu L$" src="form_4.png" width="18" height="13"/></picture>, where <picture><source srcset="form_5_dark.png" media="(prefers-color-scheme: dark)"/><img class="formulaInl" alt="$\mu$" src="form_5.png" width="9" height="9"/></picture> is the mutation rate per nucleotide site per generation, and <picture><source srcset="form_6_dark.png" media="(prefers-color-scheme: dark)"/><img class="formulaInl" alt="$L$" src="form_6.png" width="9" height="10"/></picture> is the number of nucleotide sites sequenced per haploid genome, including monomorphic sites but excluding those that fail quality control. For example, adding <span class="tt">-U 18</span> to the command above led to the following output: </p><pre class="fragment">#######################################
# legosim: site pattern probabilities #
#            version 1.89             #
#######################################

# Program was compiled: Oct 22 2020 22:30:28
# Program was run: Fri Oct 23 12:35:53 2020

# cmd: legosim -1 -i 10000 -U 18 input.lgo
# input file                  : input.lgo
# algorithm                   : stochastic
# nreps                       : 10000
# mutations per haploid genome: 18.000000
# including singleton site patterns.
#       SitePat           Count
              x               4
              y               1
              n               8
            x:y               5
            x:n               0
            y:n               0
</pre><p>Now, the 2nd column is labeled "Count" rather than "Prob" and gives the simulated count of each site pattern across the genome as a whole. It is calculated by sampling from a Poisson distribution with mean <picture><source srcset="form_7_dark.png" media="(prefers-color-scheme: dark)"/><img class="formulaInl" alt="$\mu L b$" src="form_7.png" width="23" height="13"/></picture>, where <picture><source srcset="form_5_dark.png" media="(prefers-color-scheme: dark)"/><img class="formulaInl" alt="$\mu$" src="form_5.png" width="9" height="9"/></picture> and <picture><source srcset="form_6_dark.png" media="(prefers-color-scheme: dark)"/><img class="formulaInl" alt="$L$" src="form_6.png" width="9" height="10"/></picture> are as described above, and <picture><source srcset="form_8_dark.png" media="(prefers-color-scheme: dark)"/><img class="formulaInl" alt="$b$" src="form_8.png" width="5" height="10"/></picture> is the average branch length. This Poisson model treats nucleotide sites as independent, ignoring linkage disequilibrium. The counts it provides are correct in expectation, but their variances in repeated runs of the program are too small.</p>
<h2 class="doxsection"><a class="anchor" id="autotoc_md13"></a>
Printing a summary of the network of nodes</h2>
<p>Use the <span class="tt">--network</span> option to print a summary of the network of nodes. As an example, let us use the following file, which is called <span class="tt">toy.lgo</span>: </p><pre class="fragment"># Toy .lgo file
time fixed    zero=0
twoN fixed     one=1
time free     Txyn=25000       # archaic-modern separation time
time fixed    Tn=2000          # time of Neanderthal admixture
time fixed    Txy=4000         # Africa-Eurasia separation time
twoN free   twoNn=1e3          # archaic population size
twoN free  twoNxy=1e4          # early modern population size
mixFrac free  mN=0.02          # Neanderthal admixture into y
segment x     t=zero   twoN=one    samples=1  # Africa
segment y     t=zero   twoN=one    samples=1  # Eurasia
segment n     t=Tn     twoN=twoNn  samples=1  # Neanderthal
segment y2    t=Tn     twoN=twoNxy            # pre-mig eurasia
segment xy    t=Txy    twoN=twoNxy            # early modern
segment xyn   t=Txyn   twoN=twoNn             # ancestral
mix    y  from y2 + mN * n      # y is a mixture of y2 and n
derive x  from xy               # x is child of xy
derive y2 from xy               # y2 is child of xy
derive xy from xyn              # xy is child of xyn
derive n  from xyn              # n is child of xyn
</pre><p>At the command line, type <span class="tt">legosim --network toy.lgo</span>. This produces output that ends with: </p><pre class="fragment">xyn: twoN=1000 ntrval=(25000.000000,inf)
  children: xy n
|xy: twoN=10000 ntrval=(4000.000000,25000.000000)
|  parents: xyn
|  children: x y2
||x: twoN=1 ntrval=(0.000000,4000.000000)
||  parents: xy
||y2: twoN=10000 ntrval=(2000.000000,4000.000000)
||  parents: xy
||  children: y
|||y: twoN=1 ntrval=(0.000000,2000.000000)
|||  parents: y2 n
|n: twoN=1000 ntrval=(2000.000000,25000.000000)
|  parents: xyn
|  children: y
Samples:
  0     x
  1     y
  2     n
</pre><p>The first line of output tells us that xyn is the root segment, that its population is <span class="tt">twoN=1000</span>, and that is spans an interval ranging from 2500 generations ago to infinity. The second line tells us that it has two "children", segments xy and n. The children and their children are described in the lines that follow, preceded by "|" characters to indicate the depth of nesting.</p>
<h2 class="doxsection"><a class="anchor" id="autotoc_md14"></a>
Plotting the network</h2>
<p>To make a graph of the network, use the <span class="tt">--plot &lt;f.dot&gt;</span> option, where <span class="tt">f.dot</span> is the name of the file to be written as output. If that file already exists, it will be overwritten. This output file is written in the .dot format used by <a href="https://graphviz.com">GraphViz</a> to describe directed graphs. I ran <span class="tt">legosim --plot toy.dot toy.lgo</span>, which produced toy.lgo, an output file that looks like this </p><pre class="fragment">digraph {
  ratio = 1.5;
  margin = 0;
  x [shape=square];
  y [shape=square];
  n [shape=square];
  { rank = same; x -&gt; y [style = invis]};
  { rank = same; y2 n};
  xyn -&gt; {xy n};
  xy -&gt; {x y2};
  y2 -&gt; y;
  n -&gt; y;
}
</pre><p>You can edit this file to tweak the resulting graph. In the first line, <span class="tt">ratio = 1.5</span> sets the aspect ratio&ndash;the ratio of height to width, and <span class="tt">margin = 0</span> says not to put white space around the graph. The "shape=square" lines ensure that segments with data are plotted with squares to distinguish them from segments without data. The line <span class="tt">{ rank = same; x -&gt; y [style = invis]}</span> says that x and y should have the same "rank" (vertical position) and that x should be to the left of right. Legosim emits a line like this so that all segments that end at time zero have the same rank and are plotted in the order that they appear in the .lgo file. The next line, <span class="tt">{ rank = same; y2 n}</span> says that y2 and n are in the same rank but allows GraphViz to decide how to arrange them within that rank. The remaining lines describe the arrows to be drawn between segments.</p>
<p>To turn this input into a .png graphics file, type <span class="tt">dot -Tpng toy.dot
&lt;blockquote&gt;&amp;zwj;toy.png</span>. This produces the graph shown below. </p>
<p><img src="./toy.png" alt="Graph of network in toy.lgo" class="inline"/></p>
<p>I often find it useful to edit the .dot file, adding lines to place segments in the same rank or to constrain their positions within a rank. To do this, just follow the conventions illustrated in the .dot file above.</p>
<dl class="section copyright"><dt>Copyright</dt><dd>Copyright (c) 2025, Alan R. Rogers <a href="#" onclick="location.href='mai'+'lto:'+'rog'+'er'+'s@a'+'nt'+'hro'+'.u'+'tah'+'.e'+'du'; return false;">roger<span class="obfuscator">.nosp@m.</span>s@an<span class="obfuscator">.nosp@m.</span>thro.<span class="obfuscator">.nosp@m.</span>utah<span class="obfuscator">.nosp@m.</span>.edu</a>. This file is released under the Internet Systems Consortium License, which can be found in file "LICENSE". </dd></dl>
</div></div><!-- contents -->
</div><!-- PageDoc -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated on <span class="timestamp"></span> for Legofit by&#160;<a href="https://www.doxygen.org/index.html"><img class="footer" src="doxygen.svg" width="104" height="31" alt="doxygen"/></a> 1.14.0
</small></address>
</div><!-- doc-content -->
</body>
</html>
