<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" lang="en-US">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=11"/>
<meta name="generator" content="Doxygen 1.14.0"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>Legofit: scrmpat</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<script type="text/javascript" src="clipboard.js"></script>
<script type="text/javascript" src="cookie.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr id="projectrow">
  <td id="projectalign">
   <div id="projectname">Legofit
   </div>
   <div id="projectbrief">infers population history from nucleotide site patterns.</div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.14.0 -->
<script type="text/javascript">
var searchBox = new SearchBox("searchBox", "search/",'.html');
</script>
<script type="text/javascript">
$(function() { codefold.init(); });
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
$(function() {
  initMenu('',true,false,'search.php','Search',false);
  $(function() { init_search(); });
});
</script>
<div id="main-nav"></div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<div id="MSearchResults">
<div class="SRPage">
<div id="SRIndex">
<div id="SRResults"></div>
<div class="SRStatus" id="Loading">Loading...</div>
<div class="SRStatus" id="Searching">Searching...</div>
<div class="SRStatus" id="NoMatches">No Matches</div>
</div>
</div>
</div>
</div>

</div><!-- top -->
<div id="doc-content">
<div><div class="header">
  <div class="headertitle"><div class="title">scrmpat </div></div>
</div><!--header-->
<div class="contents">
<div class="textblock"><p>Tabulate site pattern frequencies from .daf files.</p>
<h1 class="doxsection"><a class="anchor" id="autotoc_md40"></a>
Scrmpat: tabulates site patterns</h1>
<p>Scrmpat reads data generated by scrm (with option -transpose-segsites) tabulates counts of nucleotide site patterns, writing the result to standard output.</p>
<h1 class="doxsection"><a class="anchor" id="autotoc_md41"></a>
Usage</h1>
<pre class="fragment">Usage: scrmpat [options] &lt;x&gt; &lt;y&gt; ...
  where &lt;x&gt;, &lt;y&gt;, etc. are arbitrary labels, whose number and order
  must agree with that of the populations specified in the scrm command
  line (using scrm arguments -I and -eI). Labels may not include the
  character ":". Reads standard input; writes to standard
  output. Max number of input files: 32.

Options may include:
   -F or --logFixed
      log fixed sites to scrmpat.log
   -a or --logAll
      log all sites to scrmpat.log
   --version
      Print version and exit
   -h or --help
      Print this message
</pre><h1 class="doxsection"><a class="anchor" id="autotoc_md42"></a>
Example</h1>
<p><span class="tt">scrmpat</span> parses a file generated using <span class="tt">scrm</span>. The <span class="tt">scrm</span> command should include the option <span class="tt">-transpose-segsites</span>. Let us assume you have done this, that file <span class="tt">foo.scrm</span> contains the output simulated by <span class="tt">scrm</span>, and that these simulated data included genotypes referring to four populations, labeled "x", "y", "n", and "d". The <span class="tt">scrmpat</span> command would look like this: </p><pre class="fragment">scrmpat --infile foo.scrm x y n d
</pre><p><span class="tt">scrmpat</span>'s notion of a "population" differs from that of <span class="tt">scrm</span>, in that <span class="tt">scrmpat</span> treats samples of different ages as separate populations, even if they reside in the same population on the <span class="tt">scrm</span> command line. For example, consider the following <span class="tt">scrm</span> command line: </p><pre class="fragment">scrm 3 -I 2 1 1 -eI 0.5 0 1
</pre><p>This specifies three haploid samples distributed across two populations. The <span class="tt">-I</span> argument says that each population has a sample at time 0. The <span class="tt">-eI</span> argument says that, in addition, population 2 has a sample at time 0.5. All three samples would be treated as separate populations by <span class="tt">scrmpat</span>. Thus, the <span class="tt">scrmpat</span> command line should list three labels, as in "scrmpat x y z".</p>
<p>In the output, site pattern "x:y" refers to the pattern in which the derived allele is present haploid samples from "x" and "y" but not on those from other populations. The order of the command-line arguments determines the order in which labels are sorted on output. Given the command line above, we would get a site pattern labeled "x:y:d" rather than, say, "y:x:d".</p>
<p>The output looks like this: </p><pre class="fragment"># scrmpat version 1.3
# Population labels: x y n d
# Number of site patterns: 10
# Tabulated 12327755 SNPs
#       SitePat             E[count]
            x:y       340952.4592501
            x:n        46874.1307236
            x:d        46034.4670204
            y:n        55137.4236715
            y:d        43535.5248078
            n:d       231953.3372578
          x:y:n        91646.1277991
          x:y:d        88476.9619569
          x:n:d        96676.3877423
          y:n:d       100311.4411513
</pre><p>The left column lists the site patterns that occur in the data. The right column gives the expected count of each site pattern. These are not integers, because they represent averages over all possible subsamples consisting of a single haploid genome from each population.</p>
<dl class="section copyright"><dt>Copyright</dt><dd>Copyright (c) 2018, Alan R. Rogers <a href="#" onclick="location.href='mai'+'lto:'+'rog'+'er'+'s@a'+'nt'+'hro'+'.u'+'tah'+'.e'+'du'; return false;">roger<span class="obfuscator">.nosp@m.</span>s@an<span class="obfuscator">.nosp@m.</span>thro.<span class="obfuscator">.nosp@m.</span>utah<span class="obfuscator">.nosp@m.</span>.edu</a>. This file is released under the Internet Systems Consortium License, which can be found in file "LICENSE". </dd></dl>
</div></div><!-- contents -->
</div><!-- PageDoc -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated on <span class="timestamp"></span> for Legofit by&#160;<a href="https://www.doxygen.org/index.html"><img class="footer" src="doxygen.svg" width="104" height="31" alt="doxygen"/></a> 1.14.0
</small></address>
</div><!-- doc-content -->
</body>
</html>
