<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" lang="en-US">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=11"/>
<meta name="generator" content="Doxygen 1.14.0"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>Legofit: simpat</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<script type="text/javascript" src="clipboard.js"></script>
<script type="text/javascript" src="cookie.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr id="projectrow">
  <td id="projectalign">
   <div id="projectname">Legofit
   </div>
   <div id="projectbrief">infers population history from nucleotide site patterns.</div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.14.0 -->
<script type="text/javascript">
var searchBox = new SearchBox("searchBox", "search/",'.html');
</script>
<script type="text/javascript">
$(function() { codefold.init(); });
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
$(function() {
  initMenu('',true,false,'search.php','Search',false);
  $(function() { init_search(); });
});
</script>
<div id="main-nav"></div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<div id="MSearchResults">
<div class="SRPage">
<div id="SRIndex">
<div id="SRResults"></div>
<div class="SRStatus" id="Loading">Loading...</div>
<div class="SRStatus" id="Searching">Searching...</div>
<div class="SRStatus" id="NoMatches">No Matches</div>
</div>
</div>
</div>
</div>

</div><!-- top -->
<div id="doc-content">
<div><div class="header">
  <div class="headertitle"><div class="title">simpat </div></div>
</div><!--header-->
<div class="contents">
<div class="textblock"><p>Tabulate site pattern frequencies from data produced by ms2sim or msprime.</p>
<h1 class="doxsection"><a class="anchor" id="autotoc_md43"></a>
simpat: tabulates site patterns</h1>
<p>Simpat reads data generated by ms2sim or msprime, tabulates counts of nucleotide site patterns, and write the result to standard output.</p>
<h1 class="doxsection"><a class="anchor" id="autotoc_md44"></a>
Usage</h1>
<pre class="fragment">Usage: simpat [options] [inputfile]
  Reads from stdin by default. Options may include:
   -V or --version
      Print version and exit
   -h or --help
      Print this message
</pre><h1 class="doxsection"><a class="anchor" id="autotoc_md45"></a>
Example</h1>
<p><span class="tt">simpat</span> parses a file generated using <span class="tt">ms2sim</span> or <span class="tt">msprime</span>. The first few lines of input should look like this: </p><pre class="fragment">npops = 3
pop sampsize
x 2
y 1
z 1
0 0 1 0 0
0 0 1 0 0
0 1 0 0 0
</pre><p>Line 1 gives the number of populations. These need not correspond to the populations of <span class="tt">ms</span> or <span class="tt">msprime</span>. For example, <span class="tt">ms</span> might define samples from the same population at different points in time, and you might want to treat these as distinct samples within <span class="tt">simpat</span>.</p>
<p>Line 2 is a header for the 3 lines that follow.</p>
<p>In lines 3-5, the 1st column is a label for a population, and the 2nd column specifies the number of haploid samples from that population. The number of lines in this section of the input should equal <span class="tt">npops</span>, as specified on line 1. The order in which these samples are listed should correspond to the order in which they occur in the lines of data that follow.</p>
<p>All remaining lines have the same format. Each provides data for a single nucleotide site, and each consists of the same number of fields, separated by white space. The 1st field labels the current chromosome (or replicate). Each remaining field is a haploid genotype, with 0 representing the ancestral allele and 1 the derived allele. These are given in the order specified in the previous section of output (lines 3-5 in the example above). For example, our example implies that columns 2-3 refer to population x, 4 refers to y, and 5 to z.</p>
<p>In the output, site pattern "x:y" refers to the pattern in which the derived allele is present haploid samples from "x" and "y" but not on those from other populations. Here is the output of a run with 4 populations, x, y, n, and d: </p><pre class="fragment"># simpat version 1.67
# Including singleton site patterns.
# Number of site patterns: 14
# Nucleotide sites: 7298790
# Sites used: 7298790
#       SitePat             E[count]
              x      1401902.0000000
              y      1239084.0000000
              n       589582.0000000
              d       598305.0000000
            x:y      1074236.0000000
            x:n        21136.0000000
            x:d        21693.0000000
            y:n        38330.0000000
            y:d        31022.0000000
            n:d      1408276.0000000
          x:y:n        45407.0000000
          x:y:d        44519.0000000
          x:n:d       324614.0000000
          y:n:d       460684.0000000
</pre><p>The left column lists the site patterns that occur in the data. The right column gives the expected count of each site pattern. These are not necessarily integers, because they represent averages over all possible subsamples consisting of a single haploid genome from each population. They are integers here only because this simulation modeled a single haploid sample from each population.</p>
<dl class="section copyright"><dt>Copyright</dt><dd>Copyright (c) 2018, Alan R. Rogers <a href="#" onclick="location.href='mai'+'lto:'+'rog'+'er'+'s@a'+'nt'+'hro'+'.u'+'tah'+'.e'+'du'; return false;">roger<span class="obfuscator">.nosp@m.</span>s@an<span class="obfuscator">.nosp@m.</span>thro.<span class="obfuscator">.nosp@m.</span>utah<span class="obfuscator">.nosp@m.</span>.edu</a>. This file is released under the Internet Systems Consortium License, which can be found in file "LICENSE". </dd></dl>
</div></div><!-- contents -->
</div><!-- PageDoc -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated on <span class="timestamp"></span> for Legofit by&#160;<a href="https://www.doxygen.org/index.html"><img class="footer" src="doxygen.svg" width="104" height="31" alt="doxygen"/></a> 1.14.0
</small></address>
</div><!-- doc-content -->
</body>
</html>
