<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=11"/>
<meta name="generator" content="Doxygen 1.9.4"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>Legofit: sitepat</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr id="projectrow">
  <td id="projectalign">
   <div id="projectname">Legofit
   </div>
   <div id="projectbrief">infers population history from nucleotide site</div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.9.4 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
var searchBox = new SearchBox("searchBox", "search",'Search','.html');
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
/* @license-end */
</script>
<div id="main-nav"></div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

</div><!-- top -->
<div><div class="header">
  <div class="headertitle"><div class="title">sitepat </div></div>
</div><!--header-->
<div class="contents">
<div class="textblock"><p >Tabulate site pattern frequencies from .raf files.</p>
<h1><a class="anchor" id="autotoc_md39"></a>
Sitepat: tabulates site patterns</h1>
<p >Sitepat reads data in .raf format and tabulates counts of nucleotide site patterns, writing the result to standard output. Optionally, it also calculates a moving-blocks bootstrap, writing each bootstrap replicate into a separate file.</p>
<h1><a class="anchor" id="autotoc_md40"></a>
Usage</h1>
<pre class="fragment">Usage: sitepat [options] &lt;x&gt;=&lt;in_1&gt; &lt;y&gt;=&lt;in_2&gt; ... outgroup=&lt;in_K&gt;
   where &lt;x&gt; and &lt;y&gt; are arbitrary labels, and &lt;in_i&gt; are input
   files in raf format. Labels may not include the character ":".
   Final label must be "outgroup". Writes to standard output.

   If input file name ends with .gz, input is decompressed using
   gunzip. Maximum number of input files: 32 plus outgroup. Minimum is
   2 plus outgroup.

Options may include:
   -f &lt;name&gt; or --bootfile &lt;name&gt;
      Bootstrap output file basename. Def: sitepat.boot.
   -r &lt;x&gt; or --bootreps &lt;x&gt;
      # of bootstrap replicates. Def: 0
   -b &lt;x&gt; or --blocksize &lt;x&gt;
      # of SNPs per block in moving-blocks bootstrap. Def: 0.
   -1 or --singletons
      Use singleton site patterns
   -m or --logMismatch
      Log REF mismatches to sitepat.log
   -A or --logAA
      Log sites with uncallable ancestral allele
   --version
      Print version and exit
   -h or --help
      Print this message
</pre><p> Reference genomes are often useful as outgroups, but they must first be aligned to the target genome. These alignments are typically distributed in axt format. To convert from axt format to raf format, see <a class="el" href="axt2raf.html">axt2raf</a>.</p>
<h1><a class="anchor" id="autotoc_md41"></a>
Example</h1>
<p >Before running <code>sitepat</code>, use <a class="el" href="index.html#raf">raf</a> to convert the input data into raf format. To save space, you may want to compress the ".raf" files using the external utility "gzip". The names of these compressed files should end in ".gz", and <code>sitepat</code> will use an external utility (<code>gunzip</code>) to decompress them. If <code>gunzip</code> has not been installed, <code>sitepat</code> cannot read compressed input files.</p>
<p >When making bootstraps (using the <code>-r</code> or <code>--bootreps</code> options), sitepat must read the data files twice. If the files are compressed, they must be decompressed twice. It may be faster to use uncompressed input files when making bootstraps.</p>
<p >Let us assume you have generated raf files, and that directory ~/raf contains a separate raf file for each population. We want to compare 4 populations, whose .raf files are <code>yri.raf</code>, <code>ceu.raf</code>, <code>altai.raf</code>, and <code>denisova.raf</code>. Ancestral alleles are to be called using <code>chimp.raf</code> as an outgroup. The following command will do this, putting the results into <code>obs.txt</code>. </p><pre class="fragment">sitepat x=raf/yri.raf \
       y=raf/ceu.raf \
       n=raf/altai.raf \
       d=raf/denisova.raf \
       outgroup=raf/chimp.raf &gt; obs.txt
</pre><p> Here, "x", "y", "n", and "d" are labels that will be used to identify site patterns in the output. For example, site pattern "x:y" refers to the pattern in which the derived allele is present haploid samples from "x" and "y" but not on those from other populations. The order of the command-line arguments determines the order in which labels are sorted on output. Given the command line above, we would get a site pattern labeled "x:y:d" rather than, say, "y:x:d".</p>
<p >The output looks like this: </p><pre class="fragment"># Population labels:
#    x = /home/rogers/raf/yri.raf
#    y = /home/rogers/raf/ceu.raf
#    n = /home/rogers/raf/altai.raf
#    d = /home/rogers/raf/denisova.raf
# Excluding singleton site patterns.
# Number of site patterns: 10
# Tabulated 12327755 SNPs
#       SitePat             E[count]
            x:y       340952.4592501
            x:n        46874.1307236
            x:d        46034.4670204
            y:n        55137.4236715
            y:d        43535.5248078
            n:d       231953.3372578
          x:y:n        91646.1277991
          x:y:d        88476.9619569
          x:n:d        96676.3877423
          y:n:d       100311.4411513
</pre><p> The left column lists the site patterns that occur in the data. The right column gives the expected count of each site pattern. These are not integers, because they represent averages over all possible subsamples consisting of a single haploid genome from each population.</p>
<p >In the raf files used as input, chromosomes should appear in lexical order. (See <a class="el" href="index.html#sortorder">this link</a> for advice about lexical sorting.) Within each chromosome, nucleotides should appear in numerical order. There should be no duplicate (chromosome, position) pairs. Otherwise, the program aborts with an error.</p>
<p >To generate a bootstrap, use the <code>--bootreps</code> option: </p><pre class="fragment">sitepat --bootreps 50 \
       x=raf/yri.raf \
       y=raf/ceu.raf \
       n=raf/altai.raf \
       d=raf/denisova.raf \
       outgroup=raf/chimp.raf &gt; obs.txt
</pre><p> This will generate not only the primary output file, <code>obs.txt</code>, but also 50 additional files, each representing a single bootstrap replicate. The primary output file now has a bootstrap confidence interval: </p><pre class="fragment"># Population labels:
#    x = /home/rogers/raf/yri.raf
#    y = /home/rogers/raf/ceu.raf
#    n = /home/rogers/raf/altai.raf
#    d = /home/rogers/raf/denisova.raf
# Excluding singleton site patterns.
# Number of site patterns: 10
# Tabulated 12327755 SNPs
# bootstrap output file = sitepat.boot
# confidence level = 95%
#       SitePat             E[count]            low           high
            x:y       340952.4592501 338825.6604586 342406.6670816
            x:n        46874.1307236  46361.5798377  47438.1857029
            x:d        46034.4670204  45605.6588012  46631.6434277
            y:n        55137.4236715  54650.0763578  55783.7051253
            y:d        43535.5248078  43110.5119922  44234.0919024
            n:d       231953.3372578 229495.3741057 234173.6878092
          x:y:n        91646.1277991  90494.0219749  92873.4443706
          x:y:d        88476.9619569  87137.1867967  89585.8431419
          x:n:d        96676.3877423  95935.5184294  97417.6241185
          y:n:d       100311.4411513  99292.9839140 101163.3457462
</pre><p> Here, <code>low</code> and <code>high</code> are the limits of a 95% confidence interval. The bootstrap output files look like <code>sitepat.boot000</code>, <code>sitepat.boot001</code>, and so on.</p>
<dl class="section copyright"><dt>Copyright</dt><dd>Copyright (c) 2016,2019 Alan R. Rogers <a href="#" onclick="location.href='mai'+'lto:'+'rog'+'er'+'s@a'+'nt'+'hro'+'.u'+'tah'+'.e'+'du'; return false;">roger<span class="obfuscator">.nosp@m.</span>s@an<span class="obfuscator">.nosp@m.</span>thro.<span class="obfuscator">.nosp@m.</span>utah<span class="obfuscator">.nosp@m.</span>.edu</a>. This file is released under the Internet Systems Consortium License, which can be found in file "LICENSE". </dd></dl>
</div></div><!-- contents -->
</div><!-- PageDoc -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated on Tue Jun 21 2022 11:13:41 for Legofit by&#160;<a href="https://www.doxygen.org/index.html"><img class="footer" src="doxygen.svg" width="104" height="31" alt="doxygen"/></a> 1.9.4
</small></address>
</body>
</html>
